File: crates/mithril-oxide-sys/cpp/Dialect/Func/IR/FuncOps.cpp
#include "FuncOps.hpp"

#include <memory>
#include <vector>

#include <mlir/Dialect/Func/IR/FuncOps.h>
#include <mlir/IR/Location.h>
#include <mlir/IR/Types.h>
#include <rust/cxx.h>


namespace mithril_oxide_sys {

std::unique_ptr<FuncOp> FuncOp_create(
    const Location &loc,
    rust::Str name,
    const FunctionType &type,
    rust::Slice<const NamedAttribute *const> attrs,
    rust::Slice<const DictionaryAttr *const> argAttrs
)
{
    std::vector<NamedAttribute> attrs_vec;
    std::vector<DictionaryAttr> argAttrs_vec;

    for (const auto &attr : attrs)
        attrs_vec.push_back(*attr);
    for (const auto &argAttr : argAttrs)
        argAttrs_vec.push_back(*argAttr);

    return std::make_unique<FuncOp>(FuncOp::create(
        loc,
        mlir::StringRef(name.data(), name.length()),
        type,
        mlir::ArrayRef(attrs_vec.data(), attrs_vec.size()),
        mlir::ArrayRef(argAttrs_vec.data(), argAttrs_vec.size())
    ));
}

std::unique_ptr<ReturnOp> ReturnOp_create(
    const Location &loc,
    rust::Slice<const Value *const > operands
)
{
    auto builder = mlir::OpBuilder(loc->getContext());
    auto state = mlir::OperationState(loc, ReturnOp::getOperationName());
    std::vector<Value> operands_vec;

    for (const auto &val : operands)
        operands_vec.push_back(*val);

    ReturnOp::build(builder, state, operands_vec);

    ReturnOp op = ReturnOp(builder.create(state));
    return std::make_unique<ReturnOp>(op);
}

std::unique_ptr<CallOp> CallOp_create(
    const Location &loc,
    rust::Slice<const Type *const > results,
    rust::Slice<const Value *const > operands
)
{
    auto builder = mlir::OpBuilder(loc->getContext());
    auto state = mlir::OperationState(loc, CallOp::getOperationName());
    std::vector<Value> operands_vec;
    std::vector<Type> results_vec;

    for (const auto &val : operands)
        operands_vec.push_back(*val);
    for (const auto &val : results)
        results_vec.push_back(*val);

    CallOp::build(builder, state, results_vec, operands_vec);

    auto op = CallOp(builder.create(state));
    return std::make_unique<CallOp>(op);
}

} // namespace mithril_oxide_sys

File: crates/mithril-oxide-sys/src/Dialect/Func/IR/FuncOps.rs
use self::ffi::c_void;
pub use self::ffi::{CallOp, FuncOp, ReturnOp};
use crate::IR::Attributes::NamedAttribute;
use cxx::UniquePtr;
use std::fmt;

#[cxx::bridge]
pub(crate) mod ffi {

    #[namespace = "mlir"]
    unsafe extern "C++" {
        include!("mithril-oxide-sys/cpp/Dialect/Func/IR/FuncOps.hpp");

        #[namespace = "mlir::func"]
        type FuncOp;
        #[namespace = "mlir::func"]
        type ReturnOp;
        #[namespace = "mlir::func"]
        type CallOp;
        type NamedAttribute = crate::IR::Attributes::NamedAttribute;
        type Operation = crate::IR::Operation::Operation;

        #[must_use]
        pub fn getOperation(self: Pin<&mut FuncOp>) -> *mut Operation;
    }

    #[namespace = "mithril_oxide_sys"]
    unsafe extern "C++" {
        type c_void = crate::c_void;
    }

    #[namespace = "mithril_oxide_sys"]
    unsafe extern "C++" {
        include!("mithril-oxide-sys/cpp/Dialect/Func/IR/FuncOps.hpp");

        unsafe fn FuncOp_create(
            context: *const c_void,
            name: &str,
            func_type: *const c_void,
            attrs: &[*const NamedAttribute],
            // DictionaryAttr
            argAttrs: &[*const c_void],
        ) -> UniquePtr<FuncOp>;

        unsafe fn ReturnOp_create(
            context: *const c_void,
            operands: &[*const c_void],
        ) -> UniquePtr<ReturnOp>;

        unsafe fn CallOp_create(
            context: *const c_void,
            operands: &[*const c_void], // values
            results: &[*const c_void],  // types
        ) -> UniquePtr<CallOp>;
    }
}

impl ffi::FuncOp {
    #[must_use]
    /// argAttrs is a array of DictionaryAttr
    pub unsafe fn new<'a>(
        loc: *const c_void,
        name: &str,
        function_type: *const c_void,
        attrs: impl IntoIterator<Item = &'a NamedAttribute>,
        argAttrs: impl IntoIterator<Item = *const c_void>,
    ) -> UniquePtr<Self> {
        let attrs_vec = attrs.into_iter().map(|x| x as *const _).collect::<Vec<_>>();
        let argAttrs_vec = argAttrs.into_iter().collect::<Vec<_>>();

        ffi::FuncOp_create(loc, name, function_type, &attrs_vec, &argAttrs_vec)
    }
}

impl ffi::ReturnOp {
    #[must_use]
    pub unsafe fn new(
        loc: *const c_void,
        attrs: impl IntoIterator<Item = *const c_void>,
    ) -> UniquePtr<Self> {
        let operands = attrs.into_iter().collect::<Vec<_>>();

        ffi::ReturnOp_create(loc, &operands)
    }
}

impl ffi::CallOp {
    #[must_use]
    pub unsafe fn new(
        loc: *const c_void,
        // values
        operands: impl IntoIterator<Item = *const c_void>,
        // types
        results: impl IntoIterator<Item = *const c_void>,
    ) -> UniquePtr<Self> {
        let operands = operands.into_iter().collect::<Vec<_>>();
        let results = results.into_iter().collect::<Vec<_>>();

        ffi::CallOp_create(loc, &operands, &results)
    }
}

impl fmt::Debug for ffi::FuncOp {
    fn fmt(&self, f: &mut fmt::Formatter) -> fmt::Result {
        f.debug_struct("FuncOp").finish_non_exhaustive()
    }
}
